vizro-ai/examples/dashboard_ui/components.py

```diff
@@ -0,0 +1,216 @@
+
+    type: Literal["offcanvas"] = "offcanvas"
+    options: List[str]
+    value: str
+
+    def build(self):
+        """Returns the off canvas component for settings."""
+        input_groups = html.Div(
+            [
+                dbc.InputGroup(
+                    [
+                        dbc.InputGroupText("API Key"),
+                        dbc.Input(placeholder="API key", type="password", id=f"{self.id}-api-key"),
+                        html.Div(
+                            dbc.Checklist(
+                                id=f"{self.id}-api-key-toggle",
+                                options=[{"label": "", "value": False}],
+                                switch=True,
+                                inline=True,
+                            ),
+                            id="toggle-div-api-key",
+                        ),
+                    ],
+                    className="mb-3",
+                ),
+                dbc.InputGroup(
+                    [
+                        dbc.InputGroupText("API base"),
+                        dbc.Input(placeholder="(optional) API base", type="password", id=f"{self.id}-api-base"),
+                        html.Div(
+                            dbc.Checklist(
+                                id=f"{self.id}-api-base-toggle",
+                                options=[{"label": "", "value": False}],
+                                switch=True,
+                                inline=True,
+                            ),
+                            id="toggle-div-api-base",
+                        ),
+                    ],
+                    className="mb-3",
+                ),
+                dbc.InputGroup(
+                    [
+                        dbc.InputGroupText("Choose your vendor"),
+                        dbc.Select(options=self.options, value=self.value, id=f"{self.id}-dropdown"),
+                    ],
+                    className="mb-3",
+                ),
+            ],
+            className="mb-3",
+        )
+
+        offcanvas = dbc.Offcanvas(
+            id=self.id,
+            children=[
+                html.Div(
+                    children=[
+                        input_groups,
+                    ]
+                )
+            ],
+            title="Settings",
+            is_open=True,
+        )
+        return offcanvas
+
+
+class MyPage(vm.Page):
+    """Custom page."""
+
+    type: Literal["my_page"] = "my_page"
+
+    def pre_build(self):
+        """Overwriting pre_build."""
+        pass
+
+
+class Icon(vm.VizroBaseModel):
+    """Icon component for settings."""
+
+    type: Literal["icon"] = "icon"
+
+    def build(self):
+        """Returns the icon for api settings."""
+        return html.Div(
+            children=[html.Span("settings", className="material-symbols-outlined", id=self.id)],
+            className="settings-div",
+        )
+
+
+class CustomDashboard(vm.Dashboard):
```

Comment:
Possible future change: you are currently inserting extra components into the dashboard layout in two different ways: `CustomDashboard` and `app.dash.layout.children.append`. It would be nicer to just do it one way.